### DISCUSSION TITLE
Generate trans-unit id for custom export button naming?

### DISCUSSION THREAD

**Discussion by skipperTux** [kimai/kimai#2940] [Questions & Answers] [answered]
Hello. Having a custom export template and following the instructions to translate the button as described in [Export - Adding export templates](https://www.kimai.org/documentation/export.html#adding-export-templates)

> You can translate the button for your template, by adding its name to the export translation file, eg. translations/export.en.xlf.

With regards to #2912 xlf files do have `id` and `resname` now. How to generate a conflict-free id for local additions? Missing the answer to

> [which tool did you use, that creates these id and resname attributes?](https://github.com/kevinpapst/kimai2/pull/1382#discussion_r367642105)

Thanks.

**Comment by kevinpapst** [comment, marked as answer]
I use this script to generate the id 
```php
<?php

$all = glob(__DIR__.'/translations/*.xlf');

foreach ($all as $file) {
    $xml = \simplexml_load_file($file);
    if (isset($xml->file->header)) {
        unset($xml->file->header);
    }
    foreach ($xml->file->body->{'trans-unit'} as $unit) {
        $source = $unit->source;
        if (!isset($unit['resname'])) {
            $unit['resname'] = $source;
        }
        $unit['id'] = strtr(substr(base64_encode(hash('sha256', $source, true)), 0, 7), '/+', '._');
    }

    $xmlDocument = new DOMDocument('1.0');
    $xmlDocument->preserveWhiteSpace = false;
    $xmlDocument->formatOutput = true;
    $xmlDocument->loadXML($xml->asXML());

    file_put_contents($file, $xmlDocument->saveXML());
}

```

but you don't need the id for local contribution.
